Re-enable the edx.org comprehensive theme

singingwolfboy opened the pull request:
The "comprehensive theming" feature was originally merged in #8271, reverted in #9955, and is now going back in. Now it works with Django's collectstatic process, and is properly tested with microsites! For more information about how this theming feature works, read the "themes/README.rst" file in this pull request.

Companion pull requests:
* https://github.com/edx/edx-microsite/pull/95
* https://github.com/edx/configuration/pull/2461
* https://github.com/edx/edx-documentation/pull/565

This pull request starts to pull out templates and assets that are specific to edx.org, and collect them into their own comprehensive theme. Currently, this theme lives in `themes/edx.org`, but the goal is to eventually move these templates and assets out of the edx-platform repo entirely. Any usage of the `IS_EDX_DOMAIN` variable in the edx-platform codebase will eventually be replaced by this comprehensive theming feature, although this pull request only targets a few of them so far:

* the logos for LMS and Studio
* the header and footer templates
* the default profile images for student profiles

Sandboxes: 
* Open edX: https://theming.sandbox.edx.org/
* edX: https://theming-edx.sandbox.edx.org/

mtyaka commented:
@singingwolfboy The [header template](https://github.com/edx/edx-platform/blob/db/edx.org-comp-theme/lms/templates/header.html) is missing `## mako` declaration at the top. Without `## mako` it doesn't render correctly on the wiki section:

![screen shot 2015-11-23 at 10 25 20](https://cloud.githubusercontent.com/assets/32585/11333236/8dbaade2-91cc-11e5-9031-86dc3bbcb8c9.png)

mtyaka commented:
`themable-footer.html` is also missing the `## mako` declaration.

mtyaka commented:
Would it make sense to break down `header.html` into `themable-header.html` and `header.html`, just like `themable-footer/footer.html`? `header.html` and `themable-footer.html` currently look very similar, so it seems strange they don't both follow the same pattern.

singingwolfboy commented:
@mtyaka Thanks for catching the `## mako` issue! I'll fix that right away.

As for the question of `header.html` vs `themable-footer.html`, this is mentioned in the `themes/README.rst` file. The `themable-footer.html` file is an extra layer of indirection, which is only necessary for compatibility with microsites. Once comprehensive theming has enough functionality to deprecate microsites, then we will be able to remove this extra layer of indirection. `header.html` doesn't need this extra indirection, because the platform has previously used a template called `navigation.html` to populate the header of the page, so we didn't need to maintain backwards compatibility with a previous name.

Since "header" and "footer" are clearer names than "themable-header" and "themable-footer", I'd rather plan to move towards the clearer names, which means treating `themable-footer.html` as an implementation detail that theme authors don't need to care about. Does that make sense?

singingwolfboy commented:
jenkins run python

mtyaka commented:
@singingwolfboy Thanks for the info! I was thinking from a theme author's point of view - the README encourages theme authors to override `header.html`, but that file only contains code that ensures backwards-compatibility with the old stanford theming system. It feels like it would make sense to pull the backwards-compatibility code into `themable-header.html`, which would be treated as an implementation detail, just like `themable-footer.html`.

singingwolfboy commented:
@mtyaka I'm concerned that theme authors will look at the `main.html` file, see that it includes `themable-header.html`, and decide to write their themed header in a template file named `themable-header.html` to override it. I'm already concerned about that for the footer, which is why I explicitly call it out in the `themes/README.rst` file. I could explicitly call it out for both, I suppose, but that still feels like a worse solution to me.

Frankly, I think it's more likely that theme authors will start by copying an existing theme, and make edits from there. In that respect, as long as we provide a good example (like the `edx.org` theme), then it shouldn't really matter either way.

mtyaka commented:
@singingwolfboy You're right, theme authors will probably start with an existing theme, so `themable-header.html` doesn't make much sense; thanks for the explanation.

singingwolfboy commented:
jenkins run bokchoy

andy-armstrong commented:
@cahrens @explorerleslie FYI here's the PR I mentioned at stand up.

explorerleslie commented:
@singingwolfboy what's the timeline you're looking to merge this? And is there a particular part you want the TNL team to focus their review on? I'm trying to prioritize this against other work TNL is doing. cc @scottrish

singingwolfboy commented:
@explorerleslie I'm hoping to get this work into the next release of edx.org. Since the release branch is cut on Fridays, that means I'd like this to be merged by Fri Dec 4.

andy-armstrong commented:
@singingwolfboy There's something off with the spacing on the learner profile. I don't know if it was caused by your changes, but I'm not seeing it on edge. You can see here there is no spacing between the message and the footer here:

https://theming-edx.sandbox.edx.org/u/staff
![image](https://cloud.githubusercontent.com/assets/5985072/11380344/70d65c2e-92c3-11e5-97cd-cda3e5bfff5d.png)

For the same thing on edge:

https://edge.edx.org/u/andya
![image](https://cloud.githubusercontent.com/assets/5985072/11380385/a3a04390-92c3-11e5-9221-5e11452015e7.png)

Actually, I see that the same footer problem is on edx.org, so this is probably a pre-existing issue:

https://courses.edx.org/u/andywarmstrong
![image](https://cloud.githubusercontent.com/assets/5985072/11380417/e6daf484-92c3-11e5-91c6-9b8ce8409cf2.png)

nedbat commented on openedx/core/djangoapps/theming/finders.py:
```diff
@@ -0,0 +1,65 @@
+class ComprehensiveThemeFinder(BaseFinder):
+    """
+    A static files finder that searches the active comprehensive theme
+    for locate files. If the ``COMP_THEME_DIR`` setting is unset, or the
```
"for locate files"?

andy-armstrong replied:
Typo: for locate files

nedbat commented on openedx/core/djangoapps/theming/storage.py:
```diff
@@ -0,0 +1,93 @@
+        ComprehensiveThemingAwareMixin,
+        CachedFilesMixin,
+        StaticFilesStorage
+    ):  # noqa
```
Do we use `# noqa`, or is that left over from Django?

singingwolfboy replied:
`# noqa` is the equivalent of `# pylint: disable` for pep8. I was getting a pep8 error if I didn't add that comment.

nedbat replied:
There must be a form that doesn't require us to use the skin-crawling "noqa" pragma. ("noqa"? This isn't even about QA in the first place!)

nedbat commented on lms/templates/main.html:
```diff
@@ -61,7 +61,13 @@
 
   <%static:css group='style-vendor'/>
-  <%static:css group='style-main'/>
+  ## route around the overwhelmingly complicated static:css nonsense
```
Is this comment still accurate? It sounds like a punt.

singingwolfboy replied:
I don't think we're losing anything, actually. I've clarified this comment.

nedbat commented on lms/templates/themable-footer.html:
```diff
@@ -0,0 +1,14 @@
+## mako
```
There have been lots of questions about the structure of these header/footer/themable whatevers.  Would comments at the top help people understand them?

singingwolfboy replied:
Good thought. Added.

nedbat commented on openedx/core/djangoapps/theming/finders.py:
```diff
@@ -0,0 +1,65 @@
+            self.storage = None
+            return
+
+        if not isinstance(settings.COMP_THEME_DIR, basestring):
```
Just `COMP_THEME_DIR` would do here.  Also, these first seven lines are very similar to the start of `ComprehensiveThemingAwareMixin.__init__`

singingwolfboy replied:
I agree that there's a similarity -- I don't think it's worth abstracting them out into a single function. (The additional indirection causes more confusion, which outweighs the benefit of not duplicating code, in my opinion.)

Can you explain what you mean by "Just `COMP_THEME_DIR` would do here"?

nedbat replied:
You assigned the local variable "COMP_THEME_DIR" on line 22, so you can use it here.  That line also accounted for the possibility of the setting being missing (do we need to?) which this line does not.

singingwolfboy replied:
Ah, got it. Fixed.

And yes, I think it's wise to account for the possibility that the setting might be missing, since we want to support the use case of *not* having a comprehensive theme enabled at all.

nedbat commented on openedx/core/djangoapps/theming/finders.py:
```diff
@@ -0,0 +1,65 @@
+"""
+Static file finders for Django.
+https://docs.djangoproject.com/en/1.8/ref/settings/#std:setting-STATICFILES_FINDERS
+Yes, this interface is private and undocumented, but we need to access it anyway.
```
Have we asked about making this interface supported? Or what others do about the private interface?

singingwolfboy replied:
I haven't asked anyone about making this interface supported. It might be worth doing that in the future, but that is additional work that is not necessary to get this pull request merged, and even if the interface becomes supported, that won't help us until we are on a version of Django that includes that supported interface.

As far as I know, other open source projects like Django-Pipeline and Django-Require simply use this interface in the same way, despite it being unsupported.

nedbat commented on openedx/core/djangoapps/theming/finders.py:
```diff
@@ -0,0 +1,65 @@
+
+        self.storage = CachedComprehensiveThemingStorage(location=THEME_STATIC_DIR)
+
+        super(ComprehensiveThemeFinder, self).__init__(*args, **kwargs)
```
It's odd to use super at the end.  I see that BaseFinder doesn't do anything, so it might not matter where the super goes.  I also see that FileSystemFinder calls it at the end.  Is there a reason?

singingwolfboy replied:
Consistency, I guess? That's the reason why I added it.

nedbat replied:
Let's move it to the top, where super() usually goes.

singingwolfboy replied:
Done.

nedbat commented on openedx/core/djangoapps/theming/storage.py:
```diff
@@ -0,0 +1,93 @@
+            path = safe_join(self.theme_location, name)
+        except ValueError:
+            raise SuspiciousOperation("Attempted access to '%s' denied." % name)
+        return os.path.exists(os.path.normpath(path))
```
Does normpath add something here?  It should return a path that represents the same file as the input, so we shouldn't need it just to check if a file exists.

andy-armstrong replied:
I'm nervous about the number of disk accesses that will be necessary as the code tries (and usually fails) to find theme override files. Should we consider some kind of caching? Should we do performance testing to get a sense for how many such calls a typical page will make?

singingwolfboy replied:
I was trying to copy the logic that Django uses as much as possible. `themed()` is a lot like `exists()`, which is [implemented in Django here](https://github.com/django/django/blob/1.8.7/django/core/files/storage.py#L293-L294). `exists()` calls [`path()`](https://github.com/django/django/blob/1.8.7/django/core/files/storage.py#L306-L307), which calls [`safe_join()`](https://github.com/django/django/blob/1.8.7/django/utils/_os.py#L61-L86). It looks like `safe_join()` now raises `SuspiciousOperation` directly, rather than raising a `ValueError` -- this probably changed between Django 1.4 and Django 1.8. I've updated my code to just let the `SuspiciousOperation` raise. It also looks like `normpath()` is no longer called in Django's code, so I've removed it from mine, as well.

We could do some kind of performance testing, but that would delay this merge further, and at this point multiple teams are hoping to leverage this theming feature in future work. Could we do that testing after the merge?

nedbat commented on openedx/core/djangoapps/theming/storage.py:
```diff
@@ -0,0 +1,93 @@
+
+    def url(self, name, *args, **kwargs):
+        """
+        Add the theme prefix the the asset URL
```
"the the"

andy-armstrong replied:
Typo: "the the"

nedbat commented on openedx/core/djangoapps/theming/templatetags/optional_include.py:
```diff
@@ -0,0 +1,77 @@
+            return ''
+
+
+@register.tag('optional_include')
```
This is a copy of 50 lines of code from Django. I understand why you had to do it, but maybe we should put a big ackknowledgment of it here or something?  We've taken on the maintenance burden here.

I'm also torn between keeping all the behavior of the original, and stripping it down to just the parts we need.

nedbat replied:
Is there a possibility of implementing a `do_optional_include` by calling the original `do_include`, then wrapping the node returned in an `OptionalIncludeNode`, so that we don't have to duplicate the code here?

singingwolfboy replied:
There is an acknowledgement in the module docstring. I'm not sure how I can be clearer.

I'm also not sure how I can reasonably wrap an `IncludeNode` in an `OptionalIncludeNode`. What would the code look like?

nedbat replied:
Sorry i missed the comment at the top.

The code for a wrapped node is simple, not sure about how to write the 50-line function the other way:
```
class OptionalIncludeNode(Node):
    def __init__(self, include_node, *args, **kwargs):
        super(OptionalIncludeNode, self).__init__(*args, **kwargs)
        self.include_node = include_node
    def render(self, context):
        try:
            return self.include_node.render(context)
        except TemplateDoesNotExist:
            return ''
```

singingwolfboy replied:
I'm already doing essentially that for the `OptionalIncludeNode` class, except I'm subclassing rather than wrapping. Same basic technique, though.

I'm more interested in re-writing the `do_include` function, and it sounds like you don't know how to do it, either. *shrug*

andy-armstrong commented on cms/envs/aws.py:
```diff
@@ -177,6 +177,7 @@
 
 # Theme overrides
 THEME_NAME = ENV_TOKENS.get('THEME_NAME', None)
+COMP_THEME_DIR = path(ENV_TOKENS.get('COMP_THEME_DIR', COMP_THEME_DIR))
```
Nit: I don't like the name ``COMP_THEME_DIR``. It seems unnecessarily confusing and won't be clear what it means without documentation.

singingwolfboy replied:
@nedbat and I discussed this, and "comprehensive theming" was the most descriptive name that we could come up with for this feature. (It can't just be the "theming" feature, because we already have several different theming features in the codebase. :sob:) Ned said that "comprehensive theming" was too long to type all the time, and suggested that we shorten it to "comp theming", which is where this variable name came from.

Do you have a suggestion for a better, less confusing name? I'm certainly open to suggestions.

andy-armstrong commented on lms/templates/main.html:
```diff
@@ -116,16 +98,24 @@
-
-% if style_overrides_file:
-  <link rel="stylesheet" type="text/css" href="${static.url(style_overrides_file)}" />
+<% ga_acct = microsite.get_value("GOOGLE_ANALYTICS_ACCOUNT", settings.GOOGLE_ANALYTICS_ACCOUNT) %>
```
Do you happen to know why we do Google Analytics in addition to segment.io? I thought the latter was meant to include Google Analytics as necessary.

andy-armstrong replied:
It looks like it is needed by microsites... it seems a shame to have microsite-specific logic live in the main template. Isn't the goal that microsites can own their own logic. Is the problem that there isn't a 'microsite' theme that individual microsites override? I can see that this would mean we'd end up duplicating the logic into each microsite if that was the case.

andy-armstrong replied:
What happens if a microsite wants to use Segment to manage its analytics?

singingwolfboy replied:
If a microsite doesn't want this Google Analytics code to be inserted, they can simply set `GOOGLE_ANALYTICS_ACCOUNT = None` in their microsite configuration. Then, they can use any analytics system they want, including Segment, simply by inserting *that* Javascript in the templates that the microsite uses.

I agree that this logic should be removed from the main template in the future. However, we're trying to maintain backwards compatibility as much as reasonable, and this is an easy thing to keep around. In the future, as comprehensive theming gains more and more features, it will be easier to start dismantling the microsites framework and migrating existing microsites to use comprehensive theming, instead.

andy-armstrong commented on themes/README.rst:
```diff
@@ -154,3 +176,56 @@ name in the ``@import`` line.
+  the header of the ``<body>`` element, so this change makes the name more
+  accurate.
+
+* ``google_analytics.html`` has been removed. The contents of this template
```
Why was this file removed? I like the convention of having one file per vendor, as it makes it easy to find and work with. Was there a constraint that made this necessary? I don't like the idea of having one monolithic template to understand.

singingwolfboy replied:
This file was removed because it was unnecessary. I agree that the convention of having one file per vendor is nice, but that's not what we're seeing here -- microsites was simply including `google_analytics.html`, without regards to the contents of that file, and there was no option to include other files like `segment.html` or `fb_analytics.html`, etc. In fact, if you look at the companion pull request that I have in the edx-microsite repository, you'll see that the `google_analytics.html` file was already being misused -- some microsites were using it for facebook analytics, for example.

There's no reason why templates can't include other templates. If theme authors want to split up their theme into multiple files (which is a great way to improve maintainability), then they can simply import those other files from their `head-extra.html` file. That way, they can name the files based on their contents!

andy-armstrong commented on openedx/core/djangoapps/theming/core.py:
```diff
@@ -29,20 +29,25 @@ def comprehensive_theme_changes(theme_dir):
     }
 
-    templates_dir = theme_dir / "lms" / "templates"
+    if settings.PROJECT_ROOT.endswith("cms"):
```
Is there really not a cleaner way to know that we are in Studio?

singingwolfboy replied:
I'd love a clearer way to know that we're in Studio. Can you suggest one?

stvstnfrd replied:
I don't think we actually need to know "that we are in Studio", as much as we need to know "in which root to path to look in for a particular file". In this case, `PROJECT_ROOT`, aka "the directory in which we already look for project resources (`{cms/lms}/{djangoapps/templates/static}/*`)" seems like what we want to be using.

Where I take issue with this is the arbitrary mapping between `cms` and `studio`. Sure, we brand the front-end as `Studio` to users, but this isn't the front-end. Since we providing overrides to layer on top of the traditional project filesystem (`platform:lms/static`, `platform:lms/templates`), it's an oddly confusing indirection to use anything other than `cms`.

I provide an alternative here:
https://github.com/stvstnfrd/edx-platform/commit/82a8d6d8a25a97b26c082521bf5d61c35eef395c

```python
path_project = path_theme / basename(settings.PROJECT_ROOT)
path_templates = path_project / 'templates'
path_static = path_project / 'static'
path_locale = path_project / 'conf' / 'locale'
path_favicon = path_project / 'static' / 'images' / 'favicon.ico'
```

This should ensure a direct, 1-to-1 correspondence.

andy-armstrong commented on openedx/core/djangoapps/theming/finders.py:
```diff
@@ -0,0 +1,65 @@
+            raise ImproperlyConfigured("Your COMP_THEME_DIR setting must be a string")
+
+        PROJECT_ROOT = getattr(settings, "PROJECT_ROOT", "")  # pylint: disable=invalid-name
+        if PROJECT_ROOT.endswith("cms"):
```
This is the second time that this idiom has been used. I think it deserves a helper method so that we have a consistent way to detect that we're running in Studio.

singingwolfboy replied:
Where would such a helper method live? (I think it's a good idea, but I've asked myself this question several times.)

andy-armstrong commented on themes/edx.org/lms/templates/footer.html:
```diff
@@ -0,0 +1,83 @@
+<% footer = get_footer(is_secure=is_secure) %>
+<%namespace name='static' file='static_content.html'/>
+
+## WARNING: These files are specific to edx.org and are not used in installations outside of that domain. Open edX users will want to use the file "footer.html" for any changes or overrides.
```
Nit: break this line

singingwolfboy replied:
I'd rather not modify this file in this pull request. I'm simply copying the `footer-edx-v3.html` file into the edx.org theme directory. Modifications to this file can happen in a future pull request.

andy-armstrong commented:
I'm trying to convince myself that it is okay that there are no tests here. It seems like some of the new logic is fairly subtle, so it might be good to verify that theming overrides actually behave as expected. Currently it seems that the only option is manual verification that the theme is working.

andy-armstrong commented:
@singingwolfboy this is looking fantastic. It is such a major improvement over the current state of affairs, and I love how much theming logic has just been removed. I have a few questions around microsites and Google Analytics, but then this all looks good to me. I'm done with my review for now, so let me know when I should look at this again.

nedbat commented on openedx/core/djangoapps/theming/templatetags/optional_include.py:
```diff
@@ -0,0 +1,77 @@
+class OptionalIncludeNode(IncludeNode):
+    def render(self, context):
+        try:
+            super(OptionalIncludeNode, self).render(context)
```
Don't we have to return this value?  Is this code working?

singingwolfboy replied:
D'oh! Thanks for catching that. Fixed.

singingwolfboy commented:
@andy-armstrong I believe I've addressed all your comments. There actually *are* a few automated tests for comprehensive theming, which were added in #8271 -- I forgot to re-enable the ones that I disabled in #9955, so thanks for the reminder. :) I've re-enabled them now, and it seems like they're passing. (The build status is only a failure because of the quality shard reporting an issue with me asking pylint to skip the `openedx/core/djangoapps/theming/templatetags/optional_include.py` file, since it's mostly copied from Django.)

Can you do a second pass of this review?

stvstnfrd commented:
Initial code/comments here: https://github.com/edx/edx-platform/pull/10746

Higher-level remarks forthcoming..

andy-armstrong commented:
:+1: Looks great to me! Thanks for addressing my feedback @singingwolfboy.

andy-armstrong commented on lms/templates/main.html:
```diff
@@ -84,31 +94,7 @@
-    header_extra_file = microsite.get_template_path('header_extra.html')
-
-    if settings.FEATURES['IS_EDX_DOMAIN'] and not is_microsite():
-        header_file = microsite.get_template_path('navigation-edx.html')
```
@saleem-latif @mattdrayer It looks like this PR removed all usages of the file navigation-edx.html, but that file is still in the source tree. Do you see any reason not to remove it? Folks have been changing it since this merged, but seemingly for no benefit.

saleem-latif replied:
@andy-armstrong Yes, I think the template can be removed as it is no longer needed.  Usage was removed when a new theme was created named `edx.org` instead of using `IS_EDX_DOMAIN` flag. edx.org theme contains all the overrides in [`heade.html`](https://github.com/edx/edx-platform/blob/master/themes/edx.org/lms/templates/header.html) that were previously contained in `navigation-edx.html `. 
@ziafazal FYI
